pandas/core/common.py

```diff
@@ -1374,36 +1374,44 @@ def _fill_zeros(result, x, y, name, fill):
     mask the nan's from x
     """
 
-    if fill is not None:
+    if fill is None or is_float_dtype(result):
+        return result
+
+    if name.startswith(('r', '__r')):
+        x,y = y,x
 
-        if name.startswith('r'):
-            x,y = y,x
+    if not isinstance(y, np.ndarray):
+        dtype, value = _infer_dtype_from_scalar(y)
+        y = np.empty(result.shape, dtype=dtype)
+        y.fill(value)
 
+    if is_integer_dtype(y):
 
-        if not isinstance(y, np.ndarray):
-            dtype, value = _infer_dtype_from_scalar(y)
-            y = np.empty(result.shape, dtype=dtype)
-            y.fill(value)
+        if (y == 0).any():
 
-        if is_integer_dtype(y):
+            # GH 7325, mask and nans must be broadcastable (also: PR 9308)
+            # Raveling and then reshaping makes np.putmask faster
+            mask = ((y == 0) & ~np.isnan(result)).ravel()
 
-            if (y.ravel() == 0).any():
-                shape = result.shape
-                result = result.ravel().astype('float64')
+            shape = result.shape
+            result = result.ravel().astype('float64', copy=False)
```

Comment:
I think `view` is slightly different since it "can cause a reinterpretation of the bytes of memory".

    In [1]: x = np.array([1,2])

    In [2]: x.astype('float64')
    Out[2]: array([ 1.,  2.])

    In [3]: x.view('float64')
    Out[3]: array([  4.94065646e-324,   9.88131292e-324])